GCRCatalogs/register.py

```diff
@@ -1,301 +1,492 @@
+        raise ValueError("Provided class is not a subclass of *required_base_class*")
+    return subclass
+
+
+def load_catalog_from_config_dict(catalog_config):
+    """
+    Loads and returns the catalog specified in *catalog_config*.
+    """
+    return import_subclass(
+        catalog_config["subclass_name"], __package__, BaseGenericCatalog
+    )(**catalog_config)
+
+
+# Classes
+
+class RootDirManager:
+    _ROOT_DIR_SIGNAL = "^/"
+    _PATH_LIKE_KEYS = (
+        "filename",
+        "addon_filename",
+        "base_dir",
+        "root_dir",
+        "catalog_root_dir",
+        "header_file",
+        "repo",
+        "table_dir",
+    )
+    _DICT_LIST_KEYS = ("catalogs",)
+    _DESC_SITE_ENV = "DESC_GCR_SITE"
+
+    def __init__(self, site_config_path=None):
+        self._site_config_path = site_config_path
+        self._default_root_dir = None
+        self._custom_root_dir = None
+        self._site_info = self._get_site_info()
+
+        if self._site_config_path:
+            site_config = load_yaml_local(self._site_config_path)
+            for k, v in site_config.items():
+                if k in self._site_info:
+                    self._default_root_dir = v
+                    break
+
+        if not self._default_root_dir:
+            warnings.warn("Default root dir has not been set!")
+
+    def _get_site_info(self):
+        """
+        Return a string which, when executing at a recognized site with
+        well-known name, will include the name for that site
+        """
+        # First look for custom env variable
+        v = os.getenv(self._DESC_SITE_ENV)
+        if v:
+            warnings.warn("Site determined from env variable {}".format(self._DESC_SITE_ENV))
+            return v
+        return socket.getfqdn()
+
+    @property
+    def root_dir(self):
+        return self._custom_root_dir or self._default_root_dir
+
+    @root_dir.setter
+    def root_dir(self, path):
+        """
+        If 'path' is acceptable, set root dir to it
+        """
+        # os.listdir will throw exception if path is ill-formed or doesn't exist
+        try:
+            os.listdir(path)
+        except FileNotFoundError:
+            warnings.warn("root_dir has been set to non-existent path '{}'".format(path))
+        except NotADirectoryError:
+            warnings.warn("root_dir has been set to a regular file '{}'; should be directory".format(path))
+        except PermissionError:
+            warnings.warn("root_dir has been set to '{}' but you have no permission to access it".format(path))
+        except OSError as e:
+            warnings.warn("root_dir has been set to '{}' but errors may occur when you try to access it: {}".format(path, e))
+        self._custom_root_dir = path
+
+    def reset_root_dir(self):
+        self._custom_root_dir = None
+
+    def resolve_root_dir(self, config_dict, config_name=None, record=None):  # pylint: disable=unused-argument
```

Comment:
Add something about optional arguments `config_name` and `record` to header comments.  Their purpose is obscure.

Reply:
Good point. Expanded docstring to make it clear.